odoo/tools/xml_utils.py

```diff
@@ -83,3 +83,74 @@ def create_xml_node(parent_node, node_name, node_value=None):
+    :param remove_blank_nodes (bool): if True, removes leaf nodes with no text (iterative, depth-first, done after remove_blank_text)
+    :param indent_level (int): depth or level of node within root tree (use -1 to leave indentation as-is)
+    :param indent_space (str): string to use for indentation (use '' to remove all indentation)
+    :returns (etree._Element): clean node, same instance that was received (if applicable)
+    """
+    xml_node = xml_node_or_string
+
+    # Input is str/bytes: parser takes care of removing blanks/comments
+    if isinstance(xml_node, (str, bytes)):
+        parser = etree.XMLParser(remove_blank_text=remove_blank_text, remove_comments=remove_comments)
+        if isinstance(xml_node, str):
+            xml_node = xml_node.encode()  # misnomer: fromstring actually reads bytes
+        xml_node = etree.fromstring(xml_node, parser=parser)
+        remove_comments = False  # indicates job is done
+        # https://lxml.de/FAQ.html: remove_blank_text only removes some blank texts
+
+    # Process leaf nodes iteratively
+    # Depth-first, so any inner node may become a leaf too (if children are removed)
+    def leaf_iter(parent_node, node):
+        for child_node in node:
+            leaf_iter(node, child_node)
+
+        # Base condition: node is not root nor inner node
+        if parent_node is not None and not len(node):
+            if remove_blank_text and node.text is not None and not node.text.strip():
+                # node.text is None iff node has self-closing tag (text='' would create closing tag)
+                node.text = ''
+            if remove_comments and (node.tag is etree.Comment):
+                parent_node.remove(node)
+            elif remove_blank_nodes and not (node.text or ''):
+                parent_node.remove(node)
+
+    leaf_iter(None, xml_node)
+
+    # Fix indentation
+    if indent_level >= 0:
+        etree.indent(xml_node, level=indent_level, space=indent_space)
+        xml_node.tail = "\n"  # Strange behavior of indent: does not introduce line feed at tail of xml_node
+
+    return xml_node
+
+def cleanup_xml_signature(xml_node_or_string, ns_map, signature_ns="ds"):
+    """Clean up the sub-tree of the provided XML signature node.
+
+    Call BEFORE signing. Makes the format compatible with XML digital signature:
+    - Removes comments and whitespace-only text
+    - Removes line feeds from the signature_prefix:Object node (tail and text)
+    :param xml_node_or_string (etree._Element, str): XML signature node (or its string/bytes representation)
+    :param ns_map (dict): dict mapping prefixes to XML namespaces (xmlns:key="value" in XML)
+    :param signature_ns (string): name of the signature namespace
+    :returns (etree._Element): clean node, same instance that was received (if applicable, see cleanup_xml_node)
+    """
+    sig_node = cleanup_xml_node(xml_node_or_string, remove_blank_nodes=False, indent_space="")
```

Comment:
I'm not sure why you remove the spaces and blank nodes?

Reply:
Not sure either, but the signatures aren't correct otherwise.
I sraped xmlsig/xades technical documentations to find an answer, in vain.
Do you know wise people who might know ?

Reply:
The comment below "may be optional" is knowledge gathered by trial and error. I left that line of code for consistency and readability, but not necessity. The rest seems mandatory.